x-pack/plugins/security/server/routes/api/v1/authenticate.js

```diff
@@ -7,11 +7,25 @@
 import Boom from 'boom';
 import Joi from 'joi';
 import { wrapError } from '../../../lib/errors';
-import { BasicCredentials } from '../../../../server/lib/authentication/providers/basic';
 import { canRedirectRequest } from '../../../lib/can_redirect_request';
 
 export function initAuthenticateApi(server) {
 
+  server.decorate('request', 'loginAttempt', (route, options) => {
```

Comment:
Feedback on the `loginAttempt` implementation is welcome, but it's not critical at this point. The internal details of this decorator are not yet my priority.

Reply:
__question:__ I'm wondering if we want to make this a bit more generic (without hardcoding of `username` and `password` parameters) so that we can use `loginAttempt` for SAML Response payload as well?

Reply:
I think that's a good thing to consider, but I'm not sure if I'll implement it that way initially. I'll ponder it more, but my current thinking is that broader refactoring across all authentication providers should wait until we have one more two more SSO-based providers like kerberos and oauth2. SAML's working pretty well at the moment, and at least currently this loginAttempt logic only matters to providers that explicitly need to power the login form.